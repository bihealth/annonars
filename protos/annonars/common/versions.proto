syntax = "proto3";

package annonars.common.versions;

// Source that data was created from.
message CreatedFrom {
    // The name of the data source.
    string name = 1;
    // The version of the data source.
    string version = 2;
}

// Version specification.
message VersionSpec {
    // Identifier of the data.
    string identifier = 1;
    // Title of the data.
    string title = 2;
    // Creator of the data.
    string creator = 3;
    // Contributors of the data.
    repeated string contributor = 4;
    // Format of the data.
    string format = 5;
    // Date of the data.
    string date = 6;
    // Version of the data.
    string version = 7;
    // Optional genome release.
    optional string genome_release = 8;
    // Data description.
    string description = 9;
    // Data source.
    repeated string source = 10;
    // Created from information.
    repeated CreatedFrom created_from = 11;
}
