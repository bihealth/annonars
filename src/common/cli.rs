//! Common CLI code.

/// Commonly used command line arguments.
#[derive(clap::Parser, Debug, Clone)]
pub struct Args {
    /// Verbosity of the program
    #[clap(flatten)]
    pub verbose: clap_verbosity_flag::Verbosity<clap_verbosity_flag::InfoLevel>,
}

/// Output format to write.
#[derive(
    Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Default, clap::ValueEnum, strum::Display,
)]
#[strum(serialize_all = "lowercase")]
pub enum OutputFormat {
    /// JSONL format.
    #[default]
    Jsonl,
}

/// Local genome release for command line arguments.
#[derive(
    Copy,
    Clone,
    Default,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Debug,
    clap::ValueEnum,
    strum::Display,
    strum::EnumString,
    enum_map::Enum,
    serde::Serialize,
    utoipa::ToSchema,
)]
#[strum(serialize_all = "lowercase")]
pub enum GenomeRelease {
    /// GRCh37 genome release.
    #[strum(serialize = "grch37")]
    #[default]
    Grch37,
    /// GRCh38 genome release.
    #[strum(serialize = "grch38")]
    Grch38,
}

impl From<GenomeRelease> for biocommons_bioutils::assemblies::Assembly {
    fn from(val: GenomeRelease) -> Self {
        match val {
            GenomeRelease::Grch37 => biocommons_bioutils::assemblies::Assembly::Grch37p10,
            GenomeRelease::Grch38 => biocommons_bioutils::assemblies::Assembly::Grch38,
        }
    }
}

/// Construct the `indicatif` style for progress bars.
pub fn indicatif_style() -> indicatif::ProgressStyle {
    let tpl = "{spinner:.green} [{elapsed_precise}] [{wide_bar:.cyan/blue}] \
    {human_pos}/{human_len} ({per_sec})";
    indicatif::ProgressStyle::with_template(tpl)
        .unwrap()
        .with_key(
            "eta",
            |state: &indicatif::ProgressState, w: &mut dyn std::fmt::Write| {
                write!(w, "{:.1}s", state.eta().as_secs_f64()).unwrap()
            },
        )
        .progress_chars("#>-")
}

/// Construct an `indicatif` progress bar with the common style.
///
/// Also, we will enable a steady tick every 0.1s and hide in tests.
pub fn progress_bar(#[allow(unused_variables)] len: usize) -> indicatif::ProgressBar {
    #[cfg(test)]
    let pb = indicatif::ProgressBar::hidden();
    #[cfg(not(test))]
    let pb = indicatif::ProgressBar::new(len as u64).with_style(indicatif_style());
    pb.enable_steady_tick(std::time::Duration::from_millis(100));
    pb
}

/// Canonical chromosome names.
///
/// Note that the mitochondrial genome runs under two names.
pub const CANONICAL: &[&str] = &[
    "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12", "13", "14", "15", "16", "17",
    "18", "19", "20", "21", "22", "X", "Y", "M", "MT",
];

/// Make a chromosome name canonical.
pub fn canonicalize(chrom: &str) -> String {
    let chrom = chrom.strip_prefix("chr").unwrap_or(chrom);
    if chrom == "M" {
        "MT".to_string()
    } else {
        chrom.to_string()
    }
}

/// Return whether the given chromosome name is a canonical one.
///
/// The prefix `"chr"` is stripped from the name before checking.
pub fn is_canonical(chrom: &str) -> bool {
    let chrom = chrom.strip_prefix("chr").unwrap_or(chrom);
    CANONICAL.contains(&chrom)
}

/// Build windows for a given assembly.
pub fn build_genome_windows(
    assembly: biocommons_bioutils::assemblies::Assembly,
    window_size: Option<usize>,
) -> Result<Vec<(String, usize, usize)>, anyhow::Error> {
    let mut result = Vec::new();

    for seq in &biocommons_bioutils::assemblies::ASSEMBLY_INFOS[assembly].sequences {
        if is_canonical(&seq.name) {
            let window_size = window_size.unwrap_or(seq.length);
            let mut start = 0;
            let mut end = window_size;
            while start < seq.length {
                if end > seq.length {
                    end = seq.length;
                }
                result.push((seq.name.clone(), start, end));
                start = end;
                end += window_size;
            }
        }
    }

    Ok(result)
}

/// Helpers to extract chromosome name from `<release>:<chrom>` string.
pub mod extract_chrom {
    use crate::common::spdi;

    /// Get chromosome from the SPDI variant.
    ///
    /// If the optional genome release was given then it is compared to the one specified
    /// in `expected_genome_release` and stripped (comparision is case insensitive).
    pub fn from_var(
        variant: &spdi::Var,
        expected_genome_release: Option<&str>,
    ) -> Result<String, anyhow::Error> {
        if variant.sequence.contains(':') {
            let mut iter = variant.sequence.rsplitn(2, ':');
            let chromosome = iter.next().unwrap();
            if let Some(genome_release) = iter.next() {
                if let Some(expected_genome_release) = expected_genome_release {
                    if genome_release.to_lowercase() != expected_genome_release.to_lowercase() {
                        return Err(anyhow::anyhow!(
                            "genome release mismatch (lowercase): expected {}, got {}",
                            expected_genome_release,
                            genome_release
                        ));
                    }
                }
            }
            Ok(super::canonicalize(chromosome))
        } else {
            Ok(super::canonicalize(&variant.sequence))
        }
    }

    /// Get chromosome from the SPDI position.
    ///
    /// See `from_var` for details.
    pub fn from_pos(
        pos: &spdi::Pos,
        expected_genome_release: Option<&str>,
    ) -> Result<String, anyhow::Error> {
        if pos.sequence.contains(':') {
            let mut iter = pos.sequence.rsplitn(2, ':');
            let chromosome = iter.next().unwrap();
            if let Some(genome_release) = iter.next() {
                if let Some(expected_genome_release) = expected_genome_release {
                    if genome_release.to_lowercase() != expected_genome_release.to_lowercase() {
                        return Err(anyhow::anyhow!(
                            "genome release mismatch (lowercase): expected {}, got {}",
                            expected_genome_release,
                            genome_release
                        ));
                    }
                }
            }
            Ok(super::canonicalize(chromosome))
        } else {
            Ok(super::canonicalize(&pos.sequence))
        }
    }

    /// Get chromosome from the SPDI range.
    ///
    /// See `from_var` for details.
    pub fn from_range(
        range: &spdi::Range,
        expected_genome_release: Option<&str>,
    ) -> Result<String, anyhow::Error> {
        if range.sequence.contains(':') {
            let mut iter = range.sequence.rsplitn(2, ':');
            let chromosome = iter.next().unwrap();
            if let Some(genome_release) = iter.next() {
                if let Some(expected_genome_release) = expected_genome_release {
                    if genome_release.to_lowercase() != expected_genome_release.to_lowercase() {
                        return Err(anyhow::anyhow!(
                            "genome release mismatch (lowercase): expected {}, got {}",
                            expected_genome_release,
                            genome_release
                        ));
                    }
                }
            }
            Ok(super::canonicalize(chromosome))
        } else {
            Ok(super::canonicalize(&range.sequence))
        }
    }
}
