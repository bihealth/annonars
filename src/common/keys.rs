//! RocksDB keys and their encoding.

/// A chromosomal position `CHROM-POS`.
#[derive(Debug, Default, PartialEq, Eq, PartialOrd, Ord, Clone)]
pub struct Pos {
    /// Chromosome name.
    pub chrom: String,
    /// 1-based start position.
    pub pos: i32,
}

impl Pos {
    /// Create new position.
    pub fn new(chrom: String, pos: i32) -> Self {
        Self { chrom, pos }
    }

    /// Create from the given chrom/pos pair.
    pub fn from(chrom: &str, pos: i32) -> Self {
        Self {
            chrom: chrom.to_string(),
            pos,
        }
    }

    /// Normalize chrom with `chrom_name_to_key`.
    pub fn with_key_as_chrom(&self) -> Self {
        Self {
            chrom: chrom_name_to_key(&self.chrom),
            pos: self.pos,
        }
    }
}

impl From<Pos> for Vec<u8> {
    fn from(val: Pos) -> Self {
        let mut result = Vec::new();

        result.extend_from_slice(chrom_name_to_key(&val.chrom).as_bytes());
        result.extend_from_slice(&val.pos.to_be_bytes());

        result
    }
}

impl From<&[u8]> for Pos {
    fn from(value: &[u8]) -> Self {
        let chrom = chrom_key_to_name(&value[0..2]);
        let pos = i32::from_be_bytes(value[2..6].try_into().unwrap());
        Self { chrom, pos }
    }
}

impl From<super::spdi::Pos> for Pos {
    fn from(other: super::spdi::Pos) -> Self {
        Self::new(other.sequence, other.position)
    }
}

/// A chromosomal change `CHROM-POS-REF-ALT`.
#[derive(Debug, Default, PartialEq, Eq, PartialOrd, Ord, Clone)]
pub struct Var {
    /// Chromosome name.
    pub chrom: String,
    /// 1-based start position.
    pub pos: i32,
    /// Reference allele string.
    pub reference: String,
    /// Alternative allele string.
    pub alternative: String,
}

impl Var {
    /// Create new VCF-style variant.
    pub fn new(chrom: String, pos: i32, reference: String, alternative: String) -> Self {
        Self {
            chrom,
            pos,
            reference,
            alternative,
        }
    }

    /// Create from the given VCF-style variant.
    pub fn from(chrom: &str, pos: i32, reference: &str, alternative: &str) -> Self {
        Self {
            chrom: chrom.to_string(),
            pos,
            reference: reference.to_string(),
            alternative: alternative.to_string(),
        }
    }

    /// Create for all alternate alleles from the given VCF record.
    pub fn from_vcf_allele(value: &noodles_vcf::variant::RecordBuf, allele_no: usize) -> Self {
        let chrom = value.reference_sequence_name().to_string();
        let pos: usize = value
            .variant_start()
            .expect("Telomeric breakends not supported")
            .get();
        let pos = i32::try_from(pos).unwrap();
        let reference = value.reference_bases().to_string();
        Var {
            chrom,
            pos,
            reference,
            alternative: value.alternate_bases().as_ref()[allele_no].to_string(),
        }
    }
}

impl From<Var> for Vec<u8> {
    fn from(val: Var) -> Self {
        let mut result = Vec::new();

        result.extend_from_slice(chrom_name_to_key(&val.chrom).as_bytes());
        result.extend_from_slice(&val.pos.to_be_bytes());
        result.extend_from_slice(val.reference.as_bytes());
        result.push(b'>');
        result.extend_from_slice(val.alternative.as_bytes());

        result
    }
}

impl From<super::spdi::Var> for Var {
    fn from(other: super::spdi::Var) -> Self {
        Self::new(
            other.sequence,
            other.position,
            other.deletion,
            other.insertion,
        )
    }
}

/// Convert chromosome to key in RocksDB.
pub fn chrom_name_to_key(name: &str) -> String {
    let chrom = if let Some(stripped) = name.strip_prefix("chr") {
        stripped
    } else {
        name
    };
    let chrom = if chrom == "M" {
        String::from("MT")
    } else if "XY".contains(chrom) {
        format!(" {chrom}")
    } else {
        String::from(chrom)
    };
    assert!(chrom.len() <= 2, "chrom = {:?}", chrom);
    assert!(!chrom.is_empty());
    if chrom.len() == 1 {
        format!("0{chrom}")
    } else {
        chrom
    }
}

/// Convert from RocksDB chromosome key part to chromosome name.
pub fn chrom_key_to_name(key: &[u8]) -> String {
    assert!(key.len() == 2);
    if key.starts_with(b"0") || key.starts_with(b" ") {
        std::str::from_utf8(&key[1..])
            .expect("could not decode UTF-8")
            .to_string()
    } else {
        std::str::from_utf8(key)
            .expect("could not decode UTF-8")
            .to_string()
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_pos() {
        let pos = Pos::from("chr1", 123);

        insta::assert_debug_snapshot!(pos);

        let buf: Vec<u8> = pos.into();

        insta::assert_debug_snapshot!(buf);
    }

    #[test]
    fn test_var() {
        let var = Var::from("chr1", 123, "A", "T");

        insta::assert_debug_snapshot!(var);

        let buf: Vec<u8> = var.into();

        insta::assert_debug_snapshot!(buf);
    }

    #[test]
    fn test_chrom_name_to_key() {
        assert_eq!(chrom_name_to_key("chr1"), "01");
        assert_eq!(chrom_name_to_key("chr21"), "21");
        assert_eq!(chrom_name_to_key("chrX"), " X");
        assert_eq!(chrom_name_to_key("chrY"), " Y");
        assert_eq!(chrom_name_to_key("chrM"), "MT");
        assert_eq!(chrom_name_to_key("chrMT"), "MT");

        assert_eq!(chrom_name_to_key("1"), "01");
        assert_eq!(chrom_name_to_key("21"), "21");
        assert_eq!(chrom_name_to_key("X"), " X");
        assert_eq!(chrom_name_to_key("Y"), " Y");
        assert_eq!(chrom_name_to_key("M"), "MT");
        assert_eq!(chrom_name_to_key("MT"), "MT");
    }

    #[test]
    fn test_chrom_key_to_name() {
        assert_eq!(chrom_key_to_name(b"01"), "1");
        assert_eq!(chrom_key_to_name(b"21"), "21");
        assert_eq!(chrom_key_to_name(b" X"), "X");
        assert_eq!(chrom_key_to_name(b" Y"), "Y");
        assert_eq!(chrom_key_to_name(b"MT"), "MT");
    }
}
