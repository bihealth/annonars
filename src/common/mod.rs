//! Common and utility code.

use std::path::{Path, PathBuf};

pub mod cli;
pub mod keys;
pub mod noodles;
pub mod spdi;

/// The version of `annonars` package.
#[cfg(not(test))]
const VERSION: &str = env!("CARGO_PKG_VERSION");

/// This allows us to override the version to `0.0.0` in tests.
pub fn version() -> &'static str {
    #[cfg(test)]
    return "0.0.0";
    #[cfg(not(test))]
    return VERSION;
}

fn _readlink_f(path: &Path, level: u32) -> Result<PathBuf, anyhow::Error> {
    if level == 0 {
        anyhow::bail!(
            "maximal recursion depth for {:?}",
            path.as_os_str().to_string_lossy()
        )
    } else if path.is_symlink() {
        _readlink_f(&std::fs::read_link(path)?, level.saturating_sub(1))
    } else {
        Ok(path.to_path_buf())
    }
}

/// Recursively resolve the `path`.
pub fn readlink_f<P>(path: P) -> Result<PathBuf, anyhow::Error>
where
    P: AsRef<Path>,
{
    _readlink_f(path.as_ref(), 20)
}
