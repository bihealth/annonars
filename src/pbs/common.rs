//! Code generate for protobufs by `prost-build`.

/// Information about versions.
pub mod versions {
    include!(concat!(env!("OUT_DIR"), "/annonars.common.versions.rs"));
    include!(concat!(
        env!("OUT_DIR"),
        "/annonars.common.versions.serde.rs"
    ));
}
