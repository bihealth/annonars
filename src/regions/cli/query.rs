//! Query genomic regions.

use std::{io::Write, sync::Arc};

use bio::{
    bio_types::genome::AbstractInterval, data_structures::interval_tree::ArrayBackedIntervalTree,
};
use prost::Message;

use crate::common::{self, cli::extract_chrom, spdi};

/// Argument group for specifying accession or range.
#[derive(clap::Args, Debug, Clone, Default)]
#[group(required = true, multiple = false)]
pub struct ArgsQuery {
    /// Query for all variants.
    #[arg(long, group = "query")]
    pub all: bool,
    /// Query for variant with a specicic accession.
    #[arg(long, group = "query")]
    pub accession: Option<String>,
    /// Specify range to query for.
    #[arg(long, group = "query")]
    pub range: Option<spdi::Range>,
}

/// Command line arguments for `regions clingen command.
#[derive(clap::Parser, Debug, Clone, Default)]
#[command(about = "query region data stored in RocksDB", long_about = None)]
pub struct Args {
    /// Path to RocksDB directory with data.
    #[arg(long)]
    pub path_rocksdb: String,
    /// Name of the column family to import into.
    #[arg(long, default_value = "regions")]
    pub cf_name: String,
    /// Output file (default is stdout == "-").
    #[arg(long, default_value = "-")]
    pub out_file: String,
    /// Output format.
    #[arg(long, default_value = "jsonl")]
    pub out_format: common::cli::OutputFormat,

    /// Variant or position to query for.
    #[command(flatten)]
    pub query: ArgsQuery,
}

/// Meta information as read from database.
#[derive(Debug)]
pub struct Meta {
    /// Genome release of data in database.
    pub genome_release: String,
}

/// Open RocksDB given path and column family name for data and metadata.
pub fn open_rocksdb<P: AsRef<std::path::Path>>(
    path_rocksdb: P,
    cf_data: &str,
    cf_meta: &str,
) -> Result<(Arc<rocksdb::DBWithThreadMode<rocksdb::MultiThreaded>>, Meta), anyhow::Error> {
    tracing::info!("Opening RocksDB database at {} (cf={})...", path_rocksdb.as_ref().display(), &cf_data);
    let before_open = std::time::Instant::now();
    let cf_names = &[cf_meta, cf_data];
    let db = Arc::new(rocksdb::DB::open_cf_for_read_only(
        &rocksdb::Options::default(),
        common::readlink_f(&path_rocksdb)?,
        cf_names,
        true,
    )?);
    tracing::info!("  reading meta information");
    let meta = {
        let cf_meta = db.cf_handle(cf_meta).unwrap();
        let genome_release = String::from_utf8(
            db.get_cf(&cf_meta, "genome-release")?
                .ok_or_else(|| anyhow::anyhow!("missing value meta:genome-release"))?,
        )?;
        Meta { genome_release }
    };

    tracing::info!("  meta:genome-release = {}", &meta.genome_release);
    tracing::info!(
        "... opening RocksDB database took {:?}",
        before_open.elapsed()
    );

    Ok((db, meta))
}

/// Open RocksDB database from command line arguments.
pub fn open_rocksdb_from_args(
    args: &Args,
) -> Result<(Arc<rocksdb::DBWithThreadMode<rocksdb::MultiThreaded>>, Meta), anyhow::Error> {
    open_rocksdb(&args.path_rocksdb, &args.cf_name, "meta")
}

/// Enumeration for the different record types that we have.
#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Record {
    /// ClinGen dosage record.
    ClingenDosage(crate::pbs::regions::clingen::Region),
}

/// The necessary data for the tree construction.
#[derive(Debug)]
pub struct TreeData {
    /// The chromosome.
    pub chromosome: String,
    /// The start position.
    pub start: u32,
    /// The stop position.
    pub stop: u32,
}

impl Record {
    fn tree_data(&self) -> TreeData {
        match self {
            Record::ClingenDosage(record) => {
                let interval =
                    super::import::clingen::genomic_location_to_interval(&record.genomic_location)
                        .expect("could not decode genomic location");
                TreeData {
                    chromosome: interval.contig().to_string(),
                    start: interval.range().start as u32,
                    stop: interval.range().end as u32,
                }
            }
        }
    }
}

/// Write a single record to `out_writer`.
fn print_record(
    out_writer: &mut Box<dyn std::io::Write>,
    output_format: common::cli::OutputFormat,
    value: &Record,
) -> Result<(), anyhow::Error> {
    match (output_format, value) {
        (common::cli::OutputFormat::Jsonl, Record::ClingenDosage(record)) => {
            writeln!(out_writer, "{}", serde_json::to_string(record)?)?
        }
    }

    Ok(())
}

/// Decode a key/record.
fn decode_record(key: &[u8], data: &[u8]) -> Result<Record, anyhow::Error> {
    Ok(if key.starts_with(b"clingen:") {
        Record::ClingenDosage(crate::pbs::regions::clingen::Region::decode(
            &mut std::io::Cursor::new(&data),
        )?)
    } else {
        let key = std::str::from_utf8(key).unwrap_or("COULD_NOT_DECODE_KEY");
        anyhow::bail!("unknown record type from key: {}", key);
    })
}

/// Iterate all regions and print to `out_writer`.
fn print_all(
    out_writer: &mut Box<dyn std::io::Write>,
    out_format: common::cli::OutputFormat,
    db: &rocksdb::DBWithThreadMode<rocksdb::MultiThreaded>,
    cf_data: &Arc<rocksdb::BoundColumnFamily>,
) -> Result<(), anyhow::Error> {
    tracing::info!("dumping all records...");

    let mut iter = db.raw_iterator_cf(cf_data);
    iter.seek(b"");
    while iter.valid() {
        if let (Some(raw_key), Some(raw_value)) = (iter.key(), iter.value()) {
            print_record(out_writer, out_format, &decode_record(raw_key, raw_value)?)?;
            iter.next();
        } else {
            break;
        }
    }

    tracing::info!("... done dumping all records");
    Ok(())
}

/// Helper data structure that provides per-chromosome interval trees for querying.
pub struct IntervalTrees {
    /// Per-chromosome interval trees.
    trees: rustc_hash::FxHashMap<String, ArrayBackedIntervalTree<u64, Vec<u8>>>,
    /// Backing RocksDB.
    db: Arc<rocksdb::DBWithThreadMode<rocksdb::MultiThreaded>>,
    /// Name of column family with data.
    cf_data_name: String,
    /// Meta information from database.
    meta: Meta,
}

impl IntervalTrees {
    /// Construct new per-contig interval trees.
    ///
    /// This will read all records from the database and build the interval trees.
    ///
    /// # Arguments
    ///
    /// * `db` - Database to read from.
    /// * `cf_data_name` - Name of column family with data.
    /// * `meta` - Meta information from database.
    ///
    /// # Returns
    ///
    /// * `Self` - New instance.
    ///
    /// # Errors
    ///
    /// * If reading from the database fails.
    pub fn with_db(
        db: Arc<rocksdb::DBWithThreadMode<rocksdb::MultiThreaded>>,
        cf_data_name: &str,
        meta: Meta,
    ) -> Result<Self, anyhow::Error> {
        let cf_data = db.cf_handle(cf_data_name).ok_or_else(|| {
            anyhow::anyhow!("no column family with name {:?} found", cf_data_name)
        })?;
        Ok(Self {
            trees: Self::build_trees(db.clone(), cf_data.clone())?,
            db: db.clone(),
            cf_data_name: cf_data_name.to_string(),
            meta,
        })
    }

    /// Build the interval trees.
    fn build_trees(
        db: Arc<rocksdb::DBWithThreadMode<rocksdb::MultiThreaded>>,
        cf_data: Arc<rocksdb::BoundColumnFamily>,
    ) -> Result<rustc_hash::FxHashMap<String, ArrayBackedIntervalTree<u64, Vec<u8>>>, anyhow::Error>
    {
        let mut result: rustc_hash::FxHashMap<String, ArrayBackedIntervalTree<u64, Vec<u8>>> =
            rustc_hash::FxHashMap::default();

        // Obtain iterator and seek to start.
        let mut iter = db.raw_iterator_cf(&cf_data);
        iter.seek(b"");
        while iter.valid() {
            if let (Some(raw_key), Some(raw_value)) = (iter.key(), iter.value()) {
                let record = decode_record(raw_key, raw_value)
                    .map_err(|e| anyhow::anyhow!("failed to decode record: {}", e))?;
                let key = iter.key().unwrap().to_vec();
                tracing::trace!("iterator at {:?} => {:?}", &key, &record);

                let TreeData {
                    chromosome,
                    start,
                    stop,
                } = record.tree_data();

                let interval = (start as u64)..(stop as u64);
                let chrom = chromosome.strip_prefix("chr").unwrap_or(&chromosome);
                tracing::trace!("contig = {} / {:?} / {:?}", &chrom, &interval, &key);
                result
                    .entry(chrom.to_string())
                    .or_default()
                    .insert(interval, key);
                assert!(result.contains_key(chrom));

                iter.next();
            } else {
                break;
            }
        }

        result.values_mut().for_each(|tree| tree.index());

        Ok(result)
    }

    /// Query for a range.
    pub fn query(&self, range: &spdi::Range) -> Result<Vec<Record>, anyhow::Error> {
        tracing::trace!("query for {:?}", &range);
        let contig = extract_chrom::from_range(range, Some(&self.meta.genome_release))?;
        let cf_data = self.db.cf_handle(&self.cf_data_name).ok_or_else(|| {
            anyhow::anyhow!("no column family with name {:?} found", &self.cf_data_name)
        })?;
        let interval = (range.start as u64 )..(range.end as u64);
        let mut result = Vec::new();
        if let Some(tree) = self.trees.get(&contig) {
            for entry in tree.find(&interval) {
                tracing::info!("found entry: {:?}", &entry);
                if let Some(raw_value) = self.db.get_cf(&cf_data, entry.data())? {
                    result.push(decode_record(&entry.data(), &raw_value)?);
                }
            }
        } else {
            tracing::warn!("unknown contig: {:?}", &contig);
        }

        Ok(result)
    }
}

/// Implementation of `tsv query` sub command.
pub fn run(common: &common::cli::Args, args: &Args) -> Result<(), anyhow::Error> {
    tracing::info!("Starting 'region query' command");
    tracing::info!("common = {:#?}", &common);
    tracing::info!("args = {:#?}", &args);

    let (db, meta) = open_rocksdb_from_args(args)?;
    let cf_data = db.cf_handle(&args.cf_name).unwrap();

    // Obtain writer to output.
    let mut out_writer = match args.out_file.as_ref() {
        "-" => Box::new(std::io::stdout()) as Box<dyn std::io::Write>,
        out_file => {
            let path = std::path::Path::new(out_file);
            Box::new(std::fs::File::create(path).unwrap()) as Box<dyn std::io::Write>
        }
    };

    tracing::info!("Running query...");
    let before_query = std::time::Instant::now();
    if let Some(range) = args.query.range.as_ref() {
        tracing::info!("for range {:?}", &range);
        tracing::info!("Building interval trees...");
        let trees = IntervalTrees::with_db(db.clone(), &args.cf_name, meta)
            .map_err(|e| anyhow::anyhow!("failed to build interval trees: {}", e))?;
        tracing::info!("... done building interval trees");
        tracing::info!("Running query...");
        let records = trees
            .query(range)
            .map_err(|e| anyhow::anyhow!("failed to query interval trees: {}", e))?;
        for record in &records {
            print_record(&mut out_writer, args.out_format, record)?;
        }
        tracing::info!("... done running query");
    } else if let Some(accession) = args.query.accession.as_ref() {
        tracing::info!("for accession {}", accession);
        let buf = db
            .get_cf(&cf_data, accession.as_bytes())
            .map_err(|e| anyhow::anyhow!("failed to query RocksDB: {}", e))?;
        if let Some(buf) = buf {
            let record = decode_record(accession.as_bytes(), &buf)?;
            print_record(&mut out_writer, args.out_format, &record)?;
        } else {
            tracing::warn!("no record found for accession {}", accession);
        }
        tracing::info!("... done running query");
    } else if args.query.all {
        tracing::info!("for all");
        print_all(&mut out_writer, args.out_format, &db, &cf_data)?;
    } else {
        unreachable!();
    }
    tracing::info!("... done querying in {:?}", before_query.elapsed());

    tracing::info!("All done. Have a nice day!");
    Ok(())
}

#[cfg(test)]
mod test {
    use std::str::FromStr as _;
    use temp_testdir::TempDir;

    use crate::common;

    #[rstest::fixture]
    fn args_args_temp() -> (common::cli::Args, super::Args, TempDir) {
        let temp = TempDir::default();
        let common = common::cli::Args {
            verbose: clap_verbosity_flag::Verbosity::new(1, 0),
        };
        let args = super::Args {
            cf_name: String::from("regions"),
            out_file: temp.join("out").to_string_lossy().to_string(),
            out_format: common::cli::OutputFormat::Jsonl,
            ..Default::default()
        };

        (common, args, temp)
    }

    #[tracing_test::traced_test]
    #[rstest::rstest]
    #[case("tests/regions/clingen/rocksdb", "clingen")]
    fn smoke_query_region_all(
        #[case] path_rocksdb: &str,
        #[case] label: &str,
        args_args_temp: (common::cli::Args, super::Args, TempDir),
    ) -> Result<(), anyhow::Error> {
        crate::common::set_snapshot_suffix!("{}", &label);

        let (common, args, _temp) = args_args_temp;
        let args = super::Args {
            path_rocksdb: path_rocksdb.to_string(),
            query: super::ArgsQuery {
                all: true,
                ..Default::default()
            },
            ..args
        };
        super::run(&common, &args)?;
        let out_data = std::fs::read_to_string(&args.out_file)?;
        insta::assert_snapshot!(&out_data);

        Ok(())
    }

    #[tracing_test::traced_test]
    #[rstest::rstest]
    #[case("tests/regions/clingen/rocksdb", "clingen:ISCA-46733", "clingen-ISCA-46733")]
    fn smoke_query_var_accession(
        #[case] path_rocksdb: &str,
        #[case] accession: &str,
        #[case] label: &str,
        args_args_temp: (common::cli::Args, super::Args, TempDir),
    ) -> Result<(), anyhow::Error> {
        crate::common::set_snapshot_suffix!("{}", &label);

        let (common, args, _temp) = args_args_temp;
        let args = super::Args {
            path_rocksdb: path_rocksdb.to_string(),
            query: super::ArgsQuery {
                accession: Some(accession.to_string()),
                ..Default::default()
            },
            ..args
        };
        super::run(&common, &args)?;
        let out_data = std::fs::read_to_string(&args.out_file)?;
        insta::assert_snapshot!(&out_data);

        Ok(())
    }

    #[tracing_test::traced_test]
    #[rstest::rstest]
    #[case(
        "tests/regions/clingen/rocksdb",
        "chrX:47880294:47952112",
        "clingen-some"
    )]
    #[case(
        "tests/regions/clingen/rocksdb",
        "chr1:47880294:47952112",
        "clingen-none"
    )]
    fn smoke_query_range(
        #[case] path_rocksdb: &str,
        #[case] range: &str,
        #[case] label: &str,
        args_args_temp: (common::cli::Args, super::Args, TempDir),
    ) -> Result<(), anyhow::Error> {
        crate::common::set_snapshot_suffix!("{}", &label);

        let (common, args, _temp) = args_args_temp;
        let args = super::Args {
            path_rocksdb: path_rocksdb.to_string(),
            query: super::ArgsQuery {
                range: Some(crate::common::spdi::Range::from_str(range)?),
                ..Default::default()
            },
            ..args
        };
        super::run(&common, &args)?;
        let out_data = std::fs::read_to_string(&args.out_file)?;
        insta::assert_snapshot!(&out_data);

        Ok(())
    }
}
