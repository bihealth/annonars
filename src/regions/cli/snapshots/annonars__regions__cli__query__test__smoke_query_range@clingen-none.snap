---
source: src/regions/cli/query.rs
expression: "&out_data"
---

